x-pack/plugins/proxy/server/proxy.ts

```diff
@@ -0,0 +1,291 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { Agent as HTTPSAgent } from 'https';
+import { Agent as HTTPAgent } from 'http';
+import { URL } from 'url';
+
+import { Request, ResponseToolkit } from 'hapi';
+import { Observable, Subscription, from } from 'rxjs';
+import { first } from 'rxjs/operators';
+import { v4 } from 'uuid';
+import Wreck from 'wreck';
+
+import {
+  PluginInitializerContext,
+  ClusterClient,
+  Logger,
+  CoreStart,
+  CoreSetup,
+  Router,
+} from 'target/types/server';
+
+import { HttpServerSetup } from 'target/types/server/http/http_server';
+import { ProxyConfig } from './proxy_config';
+
+export enum RouteState {
+  Initializing,
+  Started,
+  Closed,
+}
+
+export interface RoutingNode {
+  type: string; // what are all the types this can be?
+  node: string;
+  state: RouteState;
+}
+
+interface LivenessNode {
+  lastUpdate: number;
+  lastChangedTime: number;
+}
+
+interface ClusterDoc {
+  nodes: {
+    [key: string]: LivenessNode;
+  };
+  routing_table: {
+    [key: string]: RoutingNode;
+  };
+}
+
+export type RoutingTable = Map<string, RoutingNode>;
+type NodeList = Map<string, LivenessNode>;
+
+// convert a map<string, V> to a plain javascript object of {[key:string]: val}
+function map2obj<V>(map: Map<string, V>): { [key: string]: V } {
+  return [...map].reduce((a, [k, v]) => Object.assign(a, { [k]: v }), {});
+}
+
+export class ProxyService {
+  private updateInterval: number = 1000 * 15;
+  private timeoutThreshold: number = 1000 * 30;
```

Comment:
probably the config should contain default values?

Reply:
Does the config have to exist?

Reply:
ah it does!